### PR TITLE
Approximate minimal edge length

### PR DESCRIPTION
**Pull request by MartinBuessemeyer**
**Description:**
Adds a file that contains the function `approximate_minimal_edge_length` which removes nodes with a degree of two if the the distance of the edge between its neighbors is smaller that a threshold and the angle between the two neighbors and the node does not exceed a threshold. 
The script accepts both nml formats: (Dictionary of `networkx` graphs and `NML`).
Additionally the `generate_nml` got a parameter that determines if id's should be globalized.
Since this PR is strongly connected to #26. I would recommend to merge #26 and then resolve merge conflicts in this branch. After that I would merge this branch into the master.

**Steps to test:**
- execute test_nml_utils_2.py with wknml as working directory

**Issue:**
- addresses #21

### CODE REVIEW

File: wknml/nml_utils_2.py [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+        current_node = graph.nodes[two_degree_node]
+        neighbors = list(graph.neighbors(two_degree_node))
+        neighbor1 = graph.nodes[neighbors[0]]
+        neighbor2 = graph.nodes[neighbors[1]]
+        vector1 = vector_between_nodes(neighbor1, current_node)
+        vector2 = vector_between_nodes(current_node, neighbor2)
+        new_edge_vector = vector_between_nodes(neighbor1, neighbor2)
+        distance_between_combined_edges = vector_length(new_edge_vector)
+        angle = calculate_angle_between_vectors(vector1, vector2)
+        if angle <= max_angle and distance_between_combined_edges <= max_length:
```
Comment (valentin-pinkau) [inline wknml/nml_utils_2.py]: This is nice!

File: wknml/nml_utils_2.py [hunk 2]
```diff
@@ -0,0 +1,77 @@
+from typing import Union, Dict, List
```
[the remark this answers is not in the document]
Comment (MartinBuessemeyer) [reply wknml/nml_utils_2.py]: Please take a look at the PR text. My plan is to have both functionality in one file.

**Comment by MartinBuessemeyer** [comment]
Changed everything to numpy function calls :)

**Comment by philippotto** [comment]
What's the state here? :) @valentin-pinkau properly needs to re-review?

**Comment by MartinBuessemeyer** [comment]
Indeed

File: wknml/nml_utils.py [hunk 3]
```diff
@@ -57,7 +47,65 @@ def detect_max_node_id_from_all_graphs(graph_dict: Dict[str, nx.Graph]) -> int:
...
+    max_id = detect_max_node_id_from_all_graphs(nml_graph)
+    next_valid_id = max_id + 1
+    for group in nml_graph.values():
+        for graph in group:
+            next_valid_id = ensure_max_edge_length_for_graph(graph, max_length, next_valid_id)
+
+    # return the same format as the input
+    if isinstance(nml_or_graph, NML):
+        return generate_nml(nml_graph, parameter_dict, globalize_ids=False)
+        return nml_graph
```
Comment (valentin-pinkau) [inline wknml/nml_utils.py]: code is unreachable

File: tests/test_utils_2.py [hunk 4]
```diff
@@ -1,5 +1,5 @@
 from wknml import parse_nml
-from wknml.nml_utils_2 import approximate_minimal_edge_length, get_vector_between_nodes, vector_length, calculate_angle_between_vectors
+from wknml.nml_utils import approximate_minimal_edge_length, get_vector_between_nodes, vector_length, calculate_angle_between_vectors
```
Comment (valentin-pinkau) [inline tests/test_utils_2.py]: This file can be deleted, right?

File: tests/test_nml_utils.py [hunk 5]
```diff
@@ -9,31 +12,59 @@ def test_ensure_max_edge_length():
...
     test_result_nml_graph = ensure_max_edge_length(test_nml_graph, max_length)
+    assert_max_length(test_result_nml_graph, max_length)
+
+    # test the nml version
+    test_result_nml = ensure_max_edge_length(test_nml, max_length)
+    test_result_nml, parameters = generate_graph(test_result_nml)
+    assert_max_length(test_result_nml, max_length)
+
 
+def assert_max_length(graph_dict: Dict[str, List[nx.Graph]], max_length: int):
```
Comment (valentin-pinkau) [inline tests/test_nml_utils.py]: max_length is float

File: wknml/nml_utils.py [hunk 6]
```diff
@@ -1,45 +1,35 @@
...
+    return np.array(node["position"])
 
-    # return the same format as the input
-    if isinstance(nml_or_graph, NML):
-        return generate_nml(nml_graph, parameter_dict, globalize=False)
-        return nml_graph
-    else:
-        return nml_graph
 
+def vector_length(vector: List[int]) -> np.ndarray:
```
Comment (valentin-pinkau) [inline wknml/nml_utils.py]: List[int] has no .dot method

File: wknml/nml_utils.py [hunk 7]
```diff
@@ -57,7 +47,65 @@ def detect_max_node_id_from_all_graphs(graph_dict: Dict[str, nx.Graph]) -> int:
...
+
+    # return the same format as the input
+    if isinstance(nml_or_graph, NML):
+        return generate_nml(nml_graph, parameter_dict, globalize_ids=False)
+        return nml_graph
+    else:
+        return nml_graph
+
+
+def ensure_max_edge_length_for_graph(graph: nx.Graph, max_length: int, current_id: int) -> int:
```
Comment (valentin-pinkau) [inline wknml/nml_utils.py]: maybe rename `previous_edge_id` to `previous_node_id`

File: tests/test_nml_utils.py [hunk 8]
```diff
@@ -9,31 +12,59 @@ def test_ensure_max_edge_length():
     # test the graph version
     test_nml_graph = generate_graph(test_nml)
     test_result_nml_graph = ensure_max_edge_length(test_nml_graph, max_length)
+    assert_max_length(test_result_nml_graph, max_length)
```
Comment (philippotto) [inline tests/test_nml_utils.py]: I think it makes sense to add another check which makes sure that the assertion is violated before running the `ensure_max_edge_length` code. Otherwise, the code would pass if the original NML already passes the assertion. However, we explicitly want to test that an NML which doesn't passes the assertion is successfully processed.

File: wknml/nml_utils.py [hunk 9]
```diff
@@ -1,45 +1,35 @@
...
-    else:
-        nml_graph = nml_or_graph[0]
-        parameter_dict = nml_or_graph[1]
-    max_id = detect_max_node_id_from_all_graphs(nml_graph)
-    next_valid_id = max_id + 1
-    for group in nml_graph.values():
-        for graph in group:
-            next_valid_id = ensure_max_edge_length_in_tree(graph, max_length, next_valid_id)
+def get_vector(node: Dict) -> np.ndarray:
+    return np.array(node["position"])
```
Comment (philippotto) [inline wknml/nml_utils.py]: I'm wondering whether this should be multiplied by the scale, since all the calculations should happen in nm space.

See https://github.com/scalableminds/wknml/issues/22
> Note that all that should happen in nm space.

also @georgwiese

File: wknml/nml_utils.py [hunk 10]
```diff
@@ -57,7 +47,64 @@ def detect_max_node_id_from_all_graphs(graph_dict: Dict[str, nx.Graph]) -> int:
...
 
 
-def ensure_max_edge_length_in_tree(graph: nx.Graph, max_length: int, current_id: int) -> int:
+def approximate_minimal_edge_length(nml_or_graph: Union[NML, nx.Graph], max_length: int, max_angle: float) -> Union[NML, nx.Graph]:
+    # it is easier to operate on a graph
+    if isinstance(nml_or_graph, NML):
+        nml_graph, parameter_dict = generate_graph(nml_or_graph)
+    else:
+        nml_graph = nml_or_graph[0]
+        parameter_dict = nml_or_graph[1]
```
Comment (philippotto) [inline wknml/nml_utils.py]: `nml_graph, parameter_dict = nml_or_graph`

File: wknml/nml_utils.py [hunk 11]
```diff
@@ -57,7 +47,64 @@ def detect_max_node_id_from_all_graphs(graph_dict: Dict[str, nx.Graph]) -> int:
...
+            approximate_minimal_edge_length_for_graph(graph, max_length, max_angle)
+
+    # return the same format as the input
+    if isinstance(nml_or_graph, NML):
+        return generate_nml(nml_graph, parameter_dict, globalize_ids=False)
+    else:
+        return nml_graph, parameter_dict
+
+
+def approximate_minimal_edge_length_for_graph(graph: nx.Graph, max_length: int, max_angle: float):
```
Comment (philippotto) [inline wknml/nml_utils.py]: I think, max_length can be a float, right?

File: wknml/nml_utils.py [hunk 12]
```diff
@@ -57,7 +47,64 @@ def detect_max_node_id_from_all_graphs(graph_dict: Dict[str, nx.Graph]) -> int:
...
+    nodes_with_degree_of_two = [node for node in graph.nodes if graph.degree(node) == 2]
+
+    for two_degree_node in nodes_with_degree_of_two:
```
Comment (philippotto) [inline wknml/nml_utils.py]: Maybe just `node_id` (instead of `two_degree_node`) which makes the difference between id and actual node clearer.

File: wknml/nml_utils.py (hunk 9 above)
Comment (MartinBuessemeyer) [reply wknml/nml_utils.py]: I don't know if the position does already include the scale since `position` is a list of 3 floats. What is the specification?

Comment (philippotto) [reply wknml/nml_utils.py]: I'd assume that the position contains voxel coordiantes (without scale), but you can simply test it, by creating a NML in wk (on some dataset where the scale is not 1,1,1) and downloading it. The positions you see in WK are voxel coordinates. If the same coordinates appear in your NML, the hypothesis is confirmed. In that case, you need to multiply the scale.

By the way, in general, it's a good idea to test your script with this workflow (create an NML with wk, add some nodes which match your distance criteria and then download, process the NML and re-upload and verify).

File: tests/test_nml_utils.py [hunk 13]
```diff
@@ -9,46 +10,57 @@ def test_ensure_max_edge_length():
...
     # test the graph interface
     test_result_nml = approximate_minimal_edge_length(test_nml, max_length, max_angle)
-    test_result_nml, parameter_dict = generate_graph(test_result_nml)
-    assert_approximate_minimal_edge_length(test_result_nml, max_length, max_angle)
+    test_result_nml, _ = generate_graph(test_result_nml)
+    assert approximate_minimal_edge_length_violated(test_result_nml, max_length, max_angle, scale)
 
 
-def assert_approximate_minimal_edge_length(graph_dict: Dict[str, List[nx.Graph]], max_length: int, max_angle: int):
+def approximate_minimal_edge_length_violated(graph_dict: Dict[str, List[nx.Graph]], max_length: int, max_angle: int, scale: np.ndarray):
```
Comment (philippotto) [inline tests/test_nml_utils.py]: How about `is_minimal_edge_length_violated`?

File: testdata/nml_with_too_long_edges.nml [hunk 14]
```diff
@@ -2,7 +2,7 @@
 <things>
   <parameters>
     <experiment name="test_dataset" />
-    <scale x="11.0" y="11.0" z="25.0" />
+    <scale x="1.0" y="1.0" z="1.0" />
```
Comment (philippotto) [inline testdata/nml_with_too_long_edges.nml]: Why did you change the scale in this (and in the other) nml file? The non-1-1-1-scale is more close to the real-world and therefore better for testing, I'd say.

Comment (MartinBuessemeyer) [reply testdata/nml_with_too_long_edges.nml]: Well when I do that, I need to change all node positions in order since all of the nodes are testing certain cases and that would be changed. Shall I rather change all node positions?

Comment (philippotto) [reply testdata/nml_with_too_long_edges.nml]: I see, yes, it would be cool if you could change the node positions instead.

Comment (MartinBuessemeyer) [reply testdata/nml_with_too_long_edges.nml]: Done ^^

File: wknml/nml_utils.py [hunk 15]
```diff
@@ -29,13 +29,13 @@ def calculate_angle_between_vectors(vector1: np.ndarray, vector2: np.ndarray) ->
 
 def calculate_distance_between_nodes(node1: Dict, node2: Dict, scale: np.ndarray) -> float:
     difference_vector = get_vector_between_nodes(node1, node2, scale)
-    return np.sqrt(difference_vector.dot(difference_vector))
+    return vector_length(difference_vector)
 
 
 def get_padding_node_position(node1: Dict, node2: Dict, relative_distance_along_vector: float, scale: np.ndarray) -> List[int]:
```
Comment (philippotto) [inline wknml/nml_utils.py]: Can you add a comment for this function? It's not entirely clear to me what the inputs/output means :see_no_evil:

Comment (philippotto) [reply wknml/nml_utils.py]: I think, you added a comment for `calculate_distance_between_nodes` but I was referring to `get_padding_node_position` :)

Comment (MartinBuessemeyer) [reply wknml/nml_utils.py]: Eh yea sry :see_no_evil:

**Comment by philippotto** [comment]
Cool stuff :+1: Feel free to merge!

File: tests/test_nml_utils.py [hunk 16]
```diff
@@ -1,39 +1,84 @@
...
-    test_result_nml, parameters = generate_graph(test_result_nml)
-    # test if all edges are smaller than the max length
-    for group in test_result_nml.values():
+    # test the graph interface
+    test_result_nml = approximate_minimal_edge_length(test_nml, max_length, max_angle)
+    test_result_nml, _ = generate_graph(test_result_nml)
+    assert is_minimal_edge_length_violated(test_result_nml, max_length, max_angle, scale)
+
+
+def is_minimal_edge_length_violated(graph_dict: Dict[str, List[nx.Graph]], max_length: int, max_angle: int, scale: np.ndarray):
```
Comment (valentin-pinkau) [inline tests/test_nml_utils.py]:
```suggestion
def is_minimal_edge_length_violated(graph_dict: Dict[str, List[nx.Graph]], max_length: float, max_angle: float, scale: np.ndarray):
```

File: wknml/nml_utils.py [hunk 17]
```diff
@@ -57,24 +51,81 @@ def detect_max_node_id_from_all_graphs(graph_dict: Dict[str, nx.Graph]) -> int:
     return max_id
 
 
-def ensure_max_edge_length_in_tree(graph: nx.Graph, max_length: int, current_id: int) -> int:
+def approximate_minimal_edge_length(nml_or_graph: Union[NML, Tuple[Dict[str, List[nx.Graph]], Dict]], max_length: int, max_angle: float) -> Union[NML, Tuple[Dict[str, List[nx.Graph]], Dict]]:
```
Comment (valentin-pinkau) [inline wknml/nml_utils.py]:
```suggestion
def approximate_minimal_edge_length(nml_or_graph: Union[NML, Tuple[Dict[str, List[nx.Graph]], Dict]], max_length: float, max_angle: float) -> Union[NML, Tuple[Dict[str, List[nx.Graph]], Dict]]:
```
